### DISCUSSION TITLE
4 Semester 5 Woche

### DISCUSSION THREAD

**Discussion by PetrusQn** [MelonigeMelone/xCalendar#32] [Blogkanal]
## Implementierung von Tests und Update auf Angular 17

Diese Woche gab es, wie sonst auch immer, enorme Fortschritte bezüglich der xCalendar-App 😛  
Wir haben uns u.A. dazu entschieden, die Angular-Version auf Angular 17 zu aktualisieren und damit einhergehend auf das _"standalone"_-Konzept der Komponenten zu setzen. Des Weiteren haben wir redundanten Programmcode entfernt und Software-Tests implementiert. Im Folgenden gehen wir auf den Test der Login-Komponente ein. 

### Test-Datei
Die Test-Datei der Komponente sieht wie folgt aus:

```
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { LoginPageComponent } from './login-page.component';
import { By } from '@angular/platform-browser';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { Router, ActivatedRoute } from '@angular/router';
import { RouterModule } from '@angular/router';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { of } from 'rxjs';

describe('LoginPageComponent', () => {
  let component: LoginPageComponent;
  let fixture: ComponentFixture<LoginPageComponent>;
  let router: Router;

  const activatedRouteStub = {
    snapshot: {
      paramMap: of({ get: () => '123' })
    }
  };

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [],
      imports: [FormsModule, ReactiveFormsModule, RouterModule, BrowserAnimationsModule],
      providers: [
        { provide: ActivatedRoute, useValue: activatedRouteStub }
      ]
    })
      .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(LoginPageComponent);
    component = fixture.componentInstance;
    router = TestBed.inject(Router);
    fixture.detectChanges();
  });

  it('should call onLogInWith with correct parameters when login button is clicked', () => {
    spyOn(component, 'onLogInWith');
    let button = fixture.debugElement.query(By.css('.login-form-login-button'));
    let userMailAddressInput = fixture.debugElement.query(By.css('#login-form-mail-address-field')).nativeElement;
    let userPasswordInput = fixture.debugElement.query(By.css('#login-form-password-field')).nativeElement;

    userMailAddressInput.value = '';
    userPasswordInput.value = '';
    userMailAddressInput.dispatchEvent(new Event('input'));
    userPasswordInput.dispatchEvent(new Event('input'));

    button.triggerEventHandler('click', null);
    expect(component.onLogInWith).toHaveBeenCalledWith('', '');
  });
});
```

### Erklärung des Tests
Ziel des Tests ist es, zu zeigen, dass die Login-Methode, mit der E-Mail-Adresse und dem Passwort aufgerufen wird, die der Benutzer in die Textfelder eingetragen hat. In obigem Beispiel, wird simuliert, dass der Benutzer nichts in die Eingabefelder eingegeben hat und anschließend wird eben überprüft, ob die ```onLogInWith``` Methode korrekt aufgerufen wird, was in diesem Fall auch so ist.

### Ausführung & Ergebnis
Der Test wird mittels Karma ausgeführt. Das Ergebnis sieht (bei Erfolg) wie folgt aus:

![image](https://github.com/MelonigeMelone/xCalendar/assets/136441045/367dd23f-1534-4fcf-b031-acc98151187a)

**Comment by MikaK21** [comment]
Hallo zusammen,
das Upgrade auf Angular 17 und die Umstellung auf das "standalone"-Konzept der Komponenten zeigen, dass ihr auf Innovation setzt, um eure App auf dem neuesten Stand zu halten.
Der Test der Login-Komponente ist gut durchdacht, indem er sicherstellt, dass die Login-Methode korrekt aufgerufen wird, wenn E-Mail und Passwort eingegeben werden. Auch finde ich es sehr gut, dass ich die Testfälle für ein leeres Eingabefeld prüft. Man sieht, dass ihr euch viele Gedanken gemacht habt. Und natürlich super, dass die Test korrekt funktionieren :)
